Ephemeral Dust

instagibbs opened the pull request:
A replacement for https://github.com/bitcoin/bitcoin/pull/29001

Now that we have 1P1C relay, TRUC transactions and sibling eviction, it makes sense to retarget this feature more narrowly by not introducing a new output type, and simple focusing on the feature of allowing temporary dust in the mempool.

Users of this can immediately use dust outputs as:
1. Single keyed anchor (can be shared by multiple parties)
2. Single unkeyed anchor, ala P2A

Which is useful when the parent transaction cannot have fees for technical or accounting reasons.

What I'm calling "keyed" anchors would be used anytime you don't want a third party to be able to run off with the utxo. As a motivating example, in Ark there is the concept of a "forfeit transaction" which spends a "connector output". The connector output would ideally be 0-value, but you would not want that utxo spend by anyone, because this would cause financial loss for the coordinator of the service: https://arkdev.info/docs/learn/concepts#forfeit-transaction

Note that this specific use-case likely doesn't work as it involves a tree of dust, but the connector idea in general demonstrates how it could be used.

Another related example is connector outputs in BitVM2: https://bitvm.org/bitvm2.html .

Note that non-TRUC usage will be impractical unless the minrelay requirement on individual transactions are dropped in general, which should happen post-cluster mempool.

Lightning Network intends to use this feature post-29.0 if available: https://github.com/lightning/bolts/issues/1171#issuecomment-2373748582 

It's also useful for Ark, ln-symmetry, spacechains, Timeout Trees, and other constructs with large presigned trees or other large-N party smart contracts.

instagibbs commented:
going to put this into draft until https://github.com/bitcoin/bitcoin/pull/29496 is merged, since dust checks are completely off when `-acceptnonstdtxn=1`. which is required for TRUC currently.

instagibbs commented:
@theStack the primary motivation is to cover cases where non-0 value is attached to handle cases where a smart contract may want to "throw away" a few sats to fees, but otherwise cannot because of the 0-fee requirement of this PR for transactions with ephemeral anchors. If the ephemeral anchor-having transaction had non-0-fee, that would allow endogenous incentives to get it mined on its own, leaving the dust in the utxo set. As an example from the LN spec, [trimmed outputs(https://github.com/lightning/bolts/blob/master/03-transactions.md#trimmed-outputs) are directly added to the commitment transaction fee. Instead, spec writers could have the value flow to the ephemeral anchor, which is then spent to fees by the child transaction.

It's a fairly narrow motivation, and honestly I don't love the timmed output to fees scheme, but also doesn't make the code anymore complicated, so I think it's worth consideration.

KristijanSajenko commented:
Thx

instagibbs commented:
rebased due to s/nVersion/version/ change for transactions causing a silent merge conflict

instagibbs commented:
rebased on master, and s/anchor/dust/ everywhere

glozow commented:
concept ACK

petertodd commented:
Concept NACK

There's no reason to limit this behavior to TRUC/V3 transactions. The implementation should not ship with this restriction.

instagibbs commented:
@petertodd I'll give a non-TRUC version another try and report back. I may have been over-thinking things.

petertodd commented:
> @petertodd I'll give a non-TRUC version another try and report back. I may have been over-thinking things. If it works out it will be very similar in code structure...
> 
> edit: oh of course I already noted in OP: We don't allow individual transactions with 0-fee outside of TRUC, for now due to trimming computation concerns. Cluster mempool does away with this restriction entirely. I'll rework the code to upgrade to non-TRUC usage once cluster mempool hits.

I think the worthwhile question here is why are 0-fee CPFP's required to signal TRUC/V3? I don't see any reason why transaction creators should be forced to adhere to the restrictive TRUC/V3 rules if they do not wish to.

t-bast commented:
Concept ACK :+1:

petertodd commented:
I'm also going to point out that restricting ephemeral dust transactions to zero fee seems to create protocol design problems around HTLCs: https://delvingbitcoin.org/t/ephemeral-anchors-and-mevil/383

The simplest solution with dust HTLCs is to just put the value towards mining fees. Forcing them to be put towards the keyless ephemeral output potentially creates incentive problems around stealing that output. If ephemeral dust transactions have fewer rules around them, that kind of design problem goes away and protocols can do what makes sense for them without trying to meet TRUC/V3 requirements.

t-bast commented:
> Forcing them to be put towards the keyless ephemeral output potentially creates incentive problems around stealing that output. 

Not really, because it will always be the miner that will steal it (since they can replace any spend by a spend to themselves). As was detailed in the delving post you're referencing, putting HTLC values in the keyless ephemeral output ensures that it's spent and doesn't pollute the utxo set, while putting the fees directly to the commitment transaction doesn't incentivize spending that output?

petertodd commented:
Stealing an ephemeral anchor output can only be done in certain conditions, at specific ranges of values. It is not always profitable due to overheads, and you can reduce those overheads by doing complex things.

The game theory is much simpler if the dust HTLC funds go directly to fees.

On July 19, 2024 5:50:49 PM GMT+02:00, Bastien Teinturier ***@***.***> wrote:
>> Forcing them to be put towards the keyless ephemeral output potentially creates incentive problems around stealing that output. 
>
>Not really, because it will always be the miner that will steal it (since they can replace any spend by a spend to themselves). As was detailed in the delving post you're referencing, putting HTLC values in the keyless ephemeral output ensures that it's spent and doesn't pollute the utxo set, while putting the fees directly to the commitment transaction doesn't incentivize spending that output?
>

instagibbs committed:
policy: Allow dust in transactions, spent in-mempool

Also known as Ephemeral Dust.

We try to ensure that dust is spent in blocks by requiring:
  - ephemeral dust tx is 0-fee
  - ephemeral dust tx only has one dust output
  - If the ephemeral dust transaction has a child,
    the dust is spent by by that child.

0-fee requirement means there is no incentive to mine
a transaction which doesn't have a child bringing its
own fees for the transaction package.

instagibbs committed:
functional test: Add ephemeral dust tests

instagibbs commented:
Ok, the rework ended up being quite easy.

Pushed an update that generalizes to non-TRUC transactions, but as noted earlier, individual transactions have to meet minrelay for now unless they are TRUC, see/divert discussion here for that specific point: https://github.com/bitcoin/bitcoin/pull/26933 . If this restriction goes away, it will be usable(see tests I added with v2 transactions at the bottom where I checked it). Note that we are still constrained to 1P1C relay until we get general package relay.

---

re:0-fee, yes it's obviously sub-optimal for transaction creation for things like trimmed values. There's no naive way that I can think of that ensures that dust isn't mined. E.g., `TxA` with dust, followed by `TxB` spending the dust. Both enter mempool. Then `TxB'` replaces `TxB`, but doesn't spend the dust output from `TxA`. If `TxA` has fees at or above minrelay, it may well be mined on its own.

I don't have any clean ways of doing that at present. If we figure it out, it's another clean relaxation that should have easy logic to follow.

@glozow had suggested long ago that we could allow any fee under minrelay for ephemeral tx, but this is still pretty limited in what amounts are allowed for fees. I'm not convinced it's worth it vs thinking harder about figuring out how to allow unbounded feerates.

achow101 commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,31 @@
+bool CheckValidEphemeralTx(const CTransaction& tx, CFeeRate dust_relay_fee, CAmount txfee, TxValidationState& state);
+
+/** Checks that all dust in package ends up spent by direct children. Assumes package is well-formed and sorted. */
+std::optional<uint256> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate);
```
In 872c06dacde6c45c2add69c8c15c94571b08119d "policy: Allow dust in transactions, spent in-mempool"

Could you document what the return value is?

instagibbs replied:
documented

achow101 commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,127 @@
+        }
+
+        if (!child_unspent_dust.empty()) {
+            return tx->GetHash();
```
In 872c06dacde6c45c2add69c8c15c94571b08119d "policy: Allow dust in transactions, spent in-mempool"

I'm a little confused about this return value. The usage of this function suggests that the return value is the wtxid of a parent transaction that has unspent dust. However, this appears to be returning the child txid.

instagibbs replied:
switched to txid since that's how it's being tracked

achow101 commented on src/validation.cpp:
```diff
@@ -1560,6 +1600,19 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    // Run package-based dust spentness checks
+    if (m_pool.m_opts.require_standard) {
+        if (const auto ephemeral_violation{CheckEphemeralSpends(txns, m_pool.m_opts.dust_relay_feerate)}) {
+            const auto parent_wtxid = ephemeral_violation.value();
```
In 872c06dacde6c45c2add69c8c15c94571b08119d "policy: Allow dust in transactions, spent in-mempool"

`CheckEphemeralSpends` appears to only return txids, not wtxid, suggest renaming this variable:

```suggestion
            const auto parent_txid = ephemeral_violation.value();
```

instagibbs replied:
done

achow101 commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,127 @@
+    std::map<Txid, uint32_t> map_tx_dust;
+
+    for (const auto& tx : package) {
+        std::unordered_set<Txid, SaltedTxidHasher> child_unspent_dust;
+        for (const auto& tx_input : tx->vin) {
+            // Parent tx had dust, child MUST be sweeping it
+            // if it's spending any output from parent
+            if (map_tx_dust.count(tx_input.prevout.hash)) {
+                child_unspent_dust.insert(tx_input.prevout.hash);
+            }
+        }
+
+        // Now that we've built a list of parent txids
+        // that have dust, make sure that all parent's
+        // dust are swept by this same tx
+        for (const auto& tx_input : tx->vin) {
+            const auto& prevout = tx_input.prevout;
+            // Parent tx had dust, child MUST be sweeping it
+            // if it's spending any output from parent
+            if (map_tx_dust.count(prevout.hash) &&
+                map_tx_dust[prevout.hash] == prevout.n) {
+                 child_unspent_dust.erase(prevout.hash);
+            }
+
+            // We want to detect dangling dust too
+            unspent_dust.erase(tx_input.prevout);
+        }
```
In 872c06dacde6c45c2add69c8c15c94571b08119d "policy: Allow dust in transactions, spent in-mempool"

It's not clear to me why there is a rule that a descendant must also spend the dust if they spent any other output from an ancestor. This rule does not seem to be mentioned anywhere in the PR description.

ISTM it would be easier and just as effective to check that all dust outputs end up being spent rather than this slightly convoluted check.

instagibbs replied:
Imagine three transactions:

TxA, 0-fee with two outputs, one non-dust, one dust
TxB, spends non-dust
TxC, spends dust

All the dust is spent if `TxA+TxB+TxC` is accepted, but the mining template may just pick up `TxA+TxB`  rather than the three "legal configurations:
0) None
1) `TxA+TxB+TxC`
2) `TxA+TxC`

By requiring the child transaction to sweep any dust from the parent txn, we ensure that there is a single child only, and this child is the only transaction possible for bringing fees, or itself being spent by another child, and so on.

instagibbs replied:
I can update the documentation reflecting this if you find the explanation compelling.

achow101 replied:
Ah, I see. Any child must spend the dust output, and I see that this is being enforced in the other variant too.

Additional documentation would be appreciated.

instagibbs replied:
added some documentation

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,127 @@
+        for (uint32_t i=0; i<tx->vout.size(); i++) {
+            if (IsDust(tx->vout[i], dust_relay_rate)) {
+                // CheckValidEphemeralTx should disallow multiples
+                Assume(map_tx_dust.count(tx->GetHash()) == 0);
```
done

instagibbs committed:
test: Add CheckMempoolEphemeralInvariants

Checks that transactions in mempool with dust
follow expected invariants.

instagibbs committed:
fuzz: add ephemeral_package_eval harness

Works a bit harder to get ephemeral dust
transactions into the mempool.

instagibbs commented:
added fuzz and unit tests which cover more than the functional tests can alone

petertodd commented:
It looks like we've created a new, rare, pinning vector here: in the event that a dust-generating transaction is reorged back into the mempool, transactions that otherwise could have spent an ordinary output of that transaction alone now can't be accepted into the mempool because they fail the "spends all parent dust" check.

For example, imagine a coinjoin is funded by a transaction A containing a dust output. After a reorg, the transaction A gets forced back into the mempool. Now the coinjoin (or double-spends of it) fails the spends all parents dust check, because it doesn't spend the dust of A, and A may never get mined due to fee-rates increasing.

instagibbs commented:
@petertodd looks like I wrote a test case for this: https://github.com/bitcoin/bitcoin/pull/30239/commits/63e097092253b25fa9d72af167d83d911c805c9c#diff-f3bb3623abbd14a7140e405175278c9187eb8c2cf6ff0279700ac507c5f4180cR334

I show that a child spend of parent tx gets into mempool even if it has dust of its own, *and* fails to spend parent's dust.

You can glance at the new usages of `bypass_limits` in the code changes to see. I should probably add a test showing that a tx with *two* dust or fee-having can't use `bypass_limits`.

instagibbs commented:
rebased on top of master to pick up PayToAnchor support for those who would like to test them together

petertodd commented:
> @petertodd I believe that's incorrect, and looks like I wrote a test case for this: [63e0970#diff-f3bb3623abbd14a7140e405175278c9187eb8c2cf6ff0279700ac507c5f4180cR334](https://github.com/bitcoin/bitcoin/commit/63e097092253b25fa9d72af167d83d911c805c9c#diff-f3bb3623abbd14a7140e405175278c9187eb8c2cf6ff0279700ac507c5f4180cR334)
> 
> I show that a child spend of parent tx gets into mempool even if it has dust of its own, _and_ fails to spend parent's dust.
> 
> Reorgs should hopefully be expensive, so we simply allow things back in mempool even if the spentness is violated.
> 
> You can glance at the new usages of `bypass_limits` in the code changes to see. I should probably add a test showing that a tx with _two_ dust or fee-having can't use `bypass_limits`.

I think you're misunderstanding the scenario I'm talking about. I'm not saying that transactions in a block won't get back into your mempool when your node reorgs the block - they of course do get it due to the `bypass_limits` flag. What I'm saying is that *subsequent* transactions won't be accepted, resulting in a pin.

Concretely, imagine that the top block contains transaction A, which contains a dust output. Transaction B, which spends an output of A, can be accepted into your mempool as A is mined, and we can double-spend B as usual with a higher fee-rate.

Now let's reorg out the top block. Transaction A is forced into the mempool using `bypass_limits`; transaction B remains in the mempool. However, we can no longer double-spend B with a higher fee-rate, as any B2 is rejected on the basis that it does not spend the dust output of A. Thus, B is pinned by the existence of A in your mempool.

While this is a relatively rare scenario, it's not *that* rare. And there's probably circumstances where attackers could take advantage of this opportunistically by waiting until they happened to have a transaction in the right state to use in an attack.

petertodd commented:
I should point out that the TRUC code may suffer from this same category of problem around reorgs too...

instagibbs commented:
Got it. Normally a reorg like this would result in an effective feerate reduction, as the entire package is now having to be paid for. With the dust spending restriction, it makes it *impossible* to replace later, exactly like a package limit pin.

> I should point out that the TRUC code may suffer from this same category of problem around reorgs too...

Depends on the topo but yep, post-reorg package limits may be hit unexpectedly. Parent->child->grandchild, where parent was formerly confirmed, as an example. Regular chain limits are also possible to hit, f.e. if the re-entered parent was 100kvB, or full chain limits hit.

Circling back to dust, I suppose bypass-ness of a transaction entry(the re-entered parent in this context) could be tracked and spenders could bypass the dust spentness check as well? Will mull it over.

petertodd commented:
> Got it. Normally a reorg like this would result in an effective feerate reduction, as the entire package is now having to be paid for. With the dust spending restriction, it makes it _impossible_ to replace later, exactly like a package limit pin.

Exactly. And unlike the inherent pin possible by the fact that the reorged transaction could simply be big, this variant works even if the transaction is small, making pulling it off potentially a lot easier.

Personally I'd consider removing the restrictions on spending outputs of ephemeral dust transactions. For what Core is trying to accomplish here, I think it's enough that the transaction was accepted into your mempool with the dust being spent. That's enough of a nudge to get the vast majority of use-cases to do what we want them to do. It's more important that this feature can be used securely by L2 implementations, without having to worry about security problems from complex edge cases.

People who want to actually create dust intentionally can do so pretty easily anyway by just sending their transactions to one of the pools that doesn't enforce the dust limit; most of the recent growth is probably due to intentional token protocols doing things for which rules like this are irrelevant.

Suppose we do remove that restriction. You would be able to create an unspent dust UTXO by broadcasting a transaction package creating the UTXO in A, a transaction with a dust output and a non-dust output, and then spending it with a two input transaction B spending both those outputs. Then you double-spend B to remove the dust input while still providing miners an incentive to mine A.

Thing is, to do this you have to spend about as much money in fees as it would have cost you in lost sats to create an unspent non-dust output instead. Probably more, because the dust limit is based on an unrealistically low 1sat/vb fee-rate. This isn't actually any worse than the status quo for an attacker trying to create dust.

Similarly, I don't think we actually need the restriction only having one dust output per transaction. If they've all been spent to get the transaction into your mempool, playing games to get them unspent costs as much or more money as just creating non-dust UTXOs.

> Regular chain limits are also possible to hit, f.e. if the re-entered parent was 100kvB, or full chain limits hit.

While I'm not sure off the top of my head if this is actually true, it *should* be the case that RBF's of existing transactions should always be allowed even if package limits have been exceeded somehow. I proposed this principle awhile back as the "always replaceable invariant": the mempool should always allow you to replace a transaction already in the mempool with a similar one at higher fee so that you can make forward progress.

This specific issue contradicts that principal because it's a hard restriction on the type of transaction allowed in a specific circumstance.

instagibbs committed:
test: unit test for CheckEphemeralSpends

instagibbs commented:
Reworked the PR to relax `IsStandardTx` to allow a single dust output, and hopefully simplified the remaining checks that enforces spend of dusty parent's dust. This should assure the same mempool invariants as before, but with more understandable and minimal changes to logic flow.

@petertodd 

> It's more important that this feature can be used securely by L2 implementations, without having to worry about security problems from complex edge cases.

I can't speak for all wallets, but at least for CLN the fresh spend of funding outputs will be detected and the anchor spending will start anew. I think for first cut the tradeoff of having invariants that are easy to reason about is best. Given that there are actionable things wallets can do and the minimal complexity, I'm sticking with the overall current approach.

> While I'm not sure off the top of my head if this is actually true, it should be the case that RBF's of existing transactions should always be allowed even if package limits have been exceeded somehow. 

Today's mempool does a suboptimal job of dealing with chain limits and RBFs. Assuming cluster mempool, we still have the issue of what to do with cluster limits that are exceeded even after removing RBF conflicts, if any. This has been called "sibling eviction" and in general seems doable in a DoS-resistant way but is left to future work. It's a problem I'm very interested in solving post-cluster mempool.

petertodd commented:
> Today's mempool does a suboptimal job of dealing with chain limits and RBFs. Assuming cluster mempool, we still have the issue of what to do with cluster limits that are exceeded even after removing RBF conflicts, if any. This has been called "sibling eviction" and in general seems doable in a DoS-resistant way but is left to future work. It's a problem I'm very interested in solving post-cluster mempool.

What I mean is, if you are doing an RBF replacement of one transaction with another transaction paying a higher fee-rate, with the same or smaller size, this should always be allowed even if somehow your mempool got into a state where a cluster limit was exceeded.

Point is, the replacement isn't making anything worse, so it should be allowed. Sibling eviction isn't needed to support this case.

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,472 @@
+        # Mempools should match via 1P1C p2p relay
+        self.num_nodes = 2
+
+    def assert_mempool_contents(self, expected=None):
+        """Assert that all transactions in expected are in the mempool,
+        and no additional ones exist.
+        """
+        if not expected:
+            expected = []
+        mempool = self.nodes[0].getrawmempool(verbose=False)
+        assert_equal(len(mempool), len(expected))
+        for tx in expected:
+            assert tx.rehash() in mempool
```
could go in test_framework/mempool_util.py?

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,472 @@
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+    def test_nonzero_dust(self):
+        self.log.info("Test that a transaction with any value is standard as long as it's spent")
```
```suggestion
        self.log.info("Test that ephemeral dust is allowed for non-0 dust values")
```

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,472 @@
+        self.connect_nodes(0, 1)
+        self.sync_all()
+
+    def test_acceptnonstd(self):
+        self.log.info("Test that dust is still allowed when acceptnonstd=1")
```
acceptnonstdtxn

instagibbs replied:
done

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,472 @@
+        res = self.nodes[0].submitpackage([dusty_tx["hex"], sweep_tx["hex"]])
+        assert_equal(res["package_msg"], "success")
+        assert_equal(len(self.nodes[0].getrawmempool()), 2)
+        self.assert_mempool_contents(expected=[dusty_tx["tx"], sweep_tx["tx"]])
```
I think all of these `assert_mempool_contents` can also have a `self.sync_mempools()` to check that the ephemeral dust packages propagate.

instagibbs replied:
added an optional sync argument, only unused during reorg test

glozow commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,126 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    // Trivial single transaction with no dust
+    BOOST_CHECK(!CheckEphemeralSpends({dust_spend}, minrelay).has_value());
+
+    // Now with dust, ok because no spending transaction exists
```
```suggestion
    // Now with dust, ok because the tx has no dusty parents
```

instagibbs replied:
done

glozow commented on src/validation.cpp:
```diff
@@ -1560,6 +1575,19 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    // Run package-based dust spentness checks
+    if (m_pool.m_opts.require_standard) {
+        if (const auto ephemeral_violation{CheckEphemeralSpends(txns, m_pool.m_opts.dust_relay_feerate)}) {
+            const Txid parent_txid = ephemeral_violation.value();
```
```suggestion
            const Txid& parent_txid = ephemeral_violation.value();
```

glozow commented on src/policy/policy.cpp:
```diff
@@ -142,11 +143,16 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
     }
 
+    // Only one dust is permitted(on otherwise valid ephemeral dust)
+    if (num_dust_outputs > 1) {
```
if you make 1 a constexpr `MAX_DUST_OUTPUTS_PER_TX`, you can `static_assert` that it == 1 as an assumption in `CheckEphemeralSpends`

instagibbs replied:
great idea, added

glozow commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,126 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    auto dust_spend_both_parents = make_tx({COutPoint{dust_txid, dust_index}, COutPoint{dust_txid_2, dust_index}}, /*version=*/2);
+    BOOST_CHECK(!CheckEphemeralSpends({grandparent_tx_1, grandparent_tx_2, dust_spend_both_parents}, minrelay).has_value());
+
+    // Spending all outputs is also correct
```
```suggestion
    // Spending other outputs is also correct, as long as the dusty one is spent
```

instagibbs replied:
done

glozow commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,117 @@
+        std::unordered_set<Txid, SaltedTxidHasher> child_unspent_dust;
+        for (const auto& tx_input : tx->vin) {
+            // Parent tx had dust, child MUST be sweeping it
+            // if it's spending any output from parent
```
Could comment that `tx_input.prevout.n` doesn't need to equal `map_tx_dust.at(tx_input.prevout.hash)`; a child can spend multiple outputs of a parent.

instagibbs replied:
elaborated a bit, let me know if it helps

glozow reviewed (commented):
Looks good. New rules seem easier to reason about (up to 1 as long as 0-fee, must spend all of my parent's dust no matter what).

Can you remind me of the use cases of a keyed dust output again? It seems slightly safer to require ephemeral dust be P2A, as it would allow anybody to clean them up in case something goes wrong, but perhaps too narrow.

Additionally, could add a release note fragment.

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,117 @@
+
+    // We never want to give incentives to mine this alone
+    if (txfee != 0) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust");
```
done

instagibbs committed:
Add release note for ephemeral dust

instagibbs commented:
With the new changes, I'm considering disallowing entry into the mempool if *either* the base fee or the modified fee is non-0. If miners really want to mine things generally with 0 fee that's one thing, but offering an interface to avoid the dust cleanup seems suboptimal.

@glozow 

> Can you remind me of the use cases of a keyed dust output again?

What I'm calling "keyed" anchors would be used anytime you don't want a third party to be able to run off with the utxo. As a motivating example, in Ark there is the concept of a "forfeit transaction" which spends a "connector output". The connector output would ideally be 0-value, but you would not want that utxo spend by anyone, because this would cause financial loss for the coordinator of the service: https://arkdev.info/docs/learn/concepts#forfeit-transaction

Note that this specific use-case likely doesn't work as it involves a tree of dust, but the connector idea in general demonstrates how it could be used.

>  It seems slightly safer to require ephemeral dust be P2A, as it would allow anybody to clean them up in case something goes wrong, but perhaps too narrow.

In general. if we have 0-value dust, I find it unlikely to hope that miners will clean it up for no reason?

I'm also pretty hesitant to guess what people will find useful in general, e.g., if they're worried about third party replacement cycling, they can be more cautious.

I feel moderately strongly about this, but of course relaxing policy further later is also another valid choice.

> Additionally, could add a release note fragment.

Added

glozow commented:
> I'm also pretty hesitant to guess what people will find useful in general, e.g., if they're worried about third party replacement cycling, they can be more cautious. The can also include a tapscript branch for anyonecanspend cleanup, just like LN does today!

Seems reasonable to want keyed. Mostly wanted the rationale to be written up here.

instagibbs commented:
@glozow copied some of the text to the OP

instagibbs commented:
CI failures seem like spurious timeouts of random tests

petertodd commented:
> In general. if we have 0-value dust, I find it unlikely to hope that miners will clean it up for no reason?

Anyone can pay fees to clean up the dust. Doesn't have to be miners specifically.

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        assert_mempool_contents(self, self.nodes[0], expected=[])
+
+    def test_fee_having_parent(self):
+        self.log.info("Test that a transaction with ephemeral dust may not have base fee, and ignores modified")
```
nit: sentence structure kind of weird. the transaction ignores modified? I know what you mean, but maybe confusing wording

instagibbs replied:
dropped the reference to modified as it probably isn't clarifying

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        sats_fee = 1
+        dusty_tx = self.wallet.create_self_transfer_multi(fee_per_output=sats_fee, version=3)
+        self.add_output_to_create_multi_result(dusty_tx)
+        assert_equal(dusty_tx["tx"].vout[0].nValue, 5000000000 - sats_fee) # main output is not dust
```
A little bit brittle, as it assumes the utxo selected is from a coinbase.
```suggestion
        assert_equal(int(COIN * dusty_tx["fee"]), sats_fee) # main output is not dust
```

instagibbs replied:
added suggestion and a direct assertion that the main output amount is >330 the taproot dust amount

glozow commented on src/validation.cpp:
```diff
@@ -934,6 +935,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
+    // Enforces 0 base fee for dust transactions, no incentive to be mined alone
+    if (m_pool.m_opts.require_standard) {
+        if (!CheckValidEphemeralTx(tx, m_pool.m_opts.dust_relay_feerate, ws.m_base_fees, state)) {
```
I can't remember if we discussed this in the past... we definitely don't want people to be able to bypass the rule using `prioritisetransaction`, but have you considered adding a `CheckValidEphemeralTx(tx, m_pool.m_opts.dust_relay_feerate, ws.m_modified_fees, state)` in addition to this check? Basically equivalent to "you can't prioritisetransaction at all," I think.

instagibbs replied:
yes, this was my open question https://github.com/bitcoin/bitcoin/pull/30239#issuecomment-2332066227

Even if we do this change it can still be prioritized for mining *after* entering the mempool; we would need heavier machinery to stop ever treating it as prioritised. This would mean someone would have to create a CPFP transaction then cycle it away, while also using `prioritisetransaction` to get it mined without the dust spent?

@sdaftuar you have any thoughts?

sdaftuar replied:
I don't have a strong view, but I'm trying to think about what a mining pool might want the behavior to be.  One guess is that some miners may want to have a way to prioritise normal-looking transactions but not inadvertently prioritise something that would create a dust output and spam the UTXO set with garbage that will never be spent.

If that is the case, then we could try to build the heavier machinery you're talking about: perhaps change the behavior of `prioritisetransaction` to produce an error by default when invoked on an in-mempool transaction that has a dust output, coupled with what I think you guys are talking about here of rejecting transactions that are prioritised already on entry.  But probably someone should talk to mining pools that offer transaction acceleration to see what they would intend and if what I just described is worth implementing; if they don't care about UTXO bloat then it's not hard to bypass these rules regardless of what we do here.

petertodd replied:
@sdaftuar 

Re: dust UTXO creation, MARA has told me directly that they do not enforce the dust limit at all, and IIUC F2Pool does not as well. I haven't however personally tested this.

instagibbs replied:
@petertodd I know for their slipstream accelerator at a minimum they do not, and perhaps they updated all their nodes to not enforce, I'm not sure.

@sdaftuar Ok that's not much machinery, and seems to make sense combined with base and modified fee both being checked.

I think any node runner who wants to ignore dust rules they'll simply set `-dustrelayfee=0`, which works out of the box as expected in this PR, and even if we're more controlling on when `prioritisetransaction` can be called.

I pushed a commit which makes the changes and can rework history if people think the changes are agreeable. Miners can evade this by simply turning dust levels to 0, like they could before.

petertodd replied:
@instagibbs What MARA told me was that their nodes do not enforce the dust limit. Notably, that means if Libre Relay or some other implementation removed the dust limit, transactions violating it would get to MARA and be mined.

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        assert_mempool_contents(self, self.nodes[0], expected=[])
+
+    def test_non_truc(self):
+        self.log.info("Test that v2 dust-having transaction is rejected even if spent, because of min relay requirement")
```
Could add a note that this is basically checking that v2 is subject to min relay feerate requirements, and you can flip this test if that changes

instagibbs replied:
dropped a note

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        self.connect_nodes(0, 1)
+        self.sync_all()
+
+    def test_acceptnonstd(self):
+        self.log.info("Test that multi-dust and non-0 fee dust is still allowed when acceptnonstdtxn=1")
+        self.restart_node(0, extra_args=["-acceptnonstdtxn=1"])
+        self.restart_node(1, extra_args=["-acceptnonstdtxn=1"])
+        self.connect_nodes(0, 1)
+
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+        # Double dust, both unspent, with fees. Would have failed individual checks.
+        dusty_tx = self.wallet.create_self_transfer_multi(fee_per_output=1000, version=3)
+        self.add_output_to_create_multi_result(dusty_tx)
+        self.add_output_to_create_multi_result(dusty_tx)
+        self.nodes[0].rpc.sendrawtransaction(dusty_tx["hex"])
+
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"]])
+
+        # Sweep tx that doesn't sweep dust ok too
+        sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=[dusty_tx["new_utxos"][0]], version=3)
+        self.nodes[0].rpc.sendrawtransaction(sweep_tx["hex"])
+
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"], sweep_tx["tx"]])
+
+        self.restart_node(0, extra_args=[])
+        self.restart_node(1, extra_args=[])
+        self.connect_nodes(0, 1)
+
+        assert_equal(self.nodes[0].getrawmempool(), [])
```
Thoughts on adding this test before the changes in commit 1?

instagibbs replied:
good idea, moved this subtest to `mempool_dust.py` as its own commit, switched it to testing dustrelay arg instead of standardness as I think that's more directly testing what we want.

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+
+        all_parent_utxos = [utxo for tx in dusty_txs for utxo in tx["new_utxos"]]
+
+        # Missing one dust spend spend from a single parent, child rejected
```
spend spend

instagibbs replied:
fixed

glozow reviewed (commented):
Finer comb of the first 2 commits

petertodd commented:
Re: dust UTXO creation, MARA has told me directly that they do not enforce the dust limit at all, and IIUC F2Pool does not as well. I haven't however personally tested this.

On September 13, 2024 3:08:05 AM GMT+03:00, Suhas Daftuar ***@***.***> wrote:
>@sdaftuar commented on this pull request.
>
>
>
>> @@ -934,6 +935,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
>                                     fSpendsCoinbase, nSigOpsCost, lock_points.value()));
>     ws.m_vsize = entry->GetTxSize();
> 
>+    // Enforces 0 base fee for dust transactions, no incentive to be mined alone
>+    if (m_pool.m_opts.require_standard) {
>+        if (!CheckValidEphemeralTx(tx, m_pool.m_opts.dust_relay_feerate, ws.m_base_fees, state)) {
>
>I don't have a strong view, but I'm trying to think about what a mining pool might want the behavior to be.  One guess is that some miners may want to have a way to prioritise normal-looking transactions but not inadvertently prioritise something that would create a dust output and spam the UTXO set with garbage that will never be spent.
>
>If that is the case, then we could try to build the heavier machinery you're talking about: perhaps change the behavior of `prioritisetransaction` to produce an error by default when invoked on an in-mempool transaction that has a dust output, coupled with what I think you guys are talking about here of rejecting transactions that are prioritised already on entry.  But probably someone should talk to mining pools that offer transaction acceleration to see what they would intend and if what I just described is worth implementing; if they don't care about UTXO bloat then it's not hard to bypass these rules regardless of what we do here.
>

instagibbs committed:
functional test: Add new -dustrelayfee=0 test case

This test would catch regressions where ephemeral
dust checks are being erroneously applied on outputs
that are not actually dust.

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        assert_equal(len(self.nodes[0].getrawmempool()), 2)
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"], sweep_tx["tx"]])
+
+        # Node restart; doesn't allow allow ephemeral transaction back in due to individual submission
```
Yes, if it's fixed this will fail, letting the author know to write a new case. Added a note that this isn't desired, just descriptive,

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,493 @@
+        # Node restart; doesn't allow allow ephemeral transaction back in due to individual submission
+        # resulting in 0-fee
+        self.restart_node(0)
+        self.restart_node(1)
```
would cause the nodes to fail to sync their mempools, no? I find it easiest when the restarts are done in sync anyways.

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,57 @@
+/** Checks that individual transactions' parents have all their dust spent by this only-child transaction.
+ *  The function returns std::nullopt if all dust is properly spent or an error message string.
+ */
+std::optional<std::string> CheckEphemeralSpends(const CTransactionRef& ptx,
```
added a sentence to each saying when they should be called

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,110 @@
+                return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust", "tx with dust output must be 0-fee");
+             }
+        }
+    }
```
taken

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,57 @@
+
+/** Does context-less checks about a single transaction.
+ * If it has relay dust, it returns false if tx has non-0 fee
+ * and sets relevant invalid state.
```
taken

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,57 @@
+
+/** These utility functions ensure that ephemeral dust is safely
+ * created and spent without risking them entering the utxo
+ * set.
```
added a hedge word

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,110 @@
+#include<policy/ephemeral_policy.h>
+#include<policy/policy.h>
+
+bool CheckValidEphemeralTx(const CTransaction& tx, CFeeRate dust_relay_fee, CAmount txfee, TxValidationState& state)
```
nope, changed

instagibbs committed:
rpc: disallow in-mempool prioritisation of dusty tx

instagibbs commented:
@fanquake can I get a 29.0 milestone for this?

sdaftuar commented on src/test/util/txmempool.cpp:
```diff
@@ -141,6 +141,54 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
     return std::nullopt;
 }
 
+std::vector<uint32_t> GetDustIndexes(const CTransactionRef tx_ref, CFeeRate dust_relay_rate)
```
Did you mean `const CTransactionRef& tx_ref` here?

instagibbs replied:
sure, changed

sdaftuar reviewed (commented):
Read through the code (except for the fuzz test, which I plan to go back and review). Concept ACK.

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,477 @@
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"], sweep_tx["tx"]])
+
+        # Doesn't spend in-mempool dust output from parent
+        unspent_sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=[dusty_tx["new_utxos"][0]], version=3)
```
done

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,477 @@
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"], sweep_tx["tx"]])
+
+        # Spend works with dust spent
+        sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=dusty_tx["new_utxos"], version=3)
```
done

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,477 @@
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+    def test_nonzero_dust(self):
+        self.log.info("Test that a single output of any size is allowed, not checking spending")
```
yes, clarified

instagibbs commented on src/validation.cpp:
```diff
@@ -1062,6 +1070,13 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     }
 
+    // Ensure any parents in-mempool that have dust have it spent by this transaction
+    if (!bypass_limits && m_pool.m_opts.require_standard) {
```
Removed the `bypass_limits` along with unification of CheckEphemeralSpends.

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,107 @@
+    // there is up to one dust output per tx.
+    static_assert(MAX_DUST_OUTPUTS_PER_TX == 1);
+
+    assert(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}));
```
changed assert to `Assume` as requested, let it stumble along

sdaftuar commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,67 @@
+#include<policy/ephemeral_policy.h>
+#include<policy/policy.h>
+
+bool CheckValidEphemeralTx(const CTransactionRef tx, CFeeRate dust_relay_fee, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
```
I think this should be `const CTransactionRef& tx` (and in the header file as well)

instagibbs replied:
done

sdaftuar commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,212 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+                                    return ProcessNewPackage(chainstate, tx_pool, txs, /*test_accept=*/single_submit, /*client_maxfeerate=*/{}));
+
+        const auto res = WITH_LOCK(::cs_main, return AcceptToMemoryPool(chainstate, txs.back(), GetTime(),
+                                   /*bypass_limits=*/false, /*test_accept=*/!single_submit));
```
Perhaps we could let `bypass_limits` be filled in by the fuzzer, rather than hardcoded to false?

instagibbs replied:
done, previously this would have allowed non-dust-spending children and invalidate the invariant checks, now it shouldn't matter

sdaftuar commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,67 @@
+                    }
+                }
+            }
+        }
```
Overall I think this is much easier to follow with the codepaths merged, thanks.  However on further review, I am concerned about how much work we might do in this function.

As written, if a transaction had 1000 inputs, which all spend from a single parent with 1000 outputs, then we'd do 1M loop iterations, because we're not deduplicating parents for each input of a transaction (so for each input of the child, we'd be looking at all 1000 outputs of the parent).  If we deduplicate the parents first, then the number of parent outputs we might possibly have to look at is bounded by the ancestor size limit (or the cluster size limit, in the future), to something on the order of a few thousand outputs. 

If we are relying on the ancestor size limits for a work bound, then we should make sure that all those limits are checked prior to this function being invoked.  I think in the package acceptance case, that would involve moving this check to come after the `CheckPackageLimits()` check that happens in `AcceptMultipleTransactions`.

I think this might be enough, but we could consider going further -- what if we were to cache how many dust outputs a transaction has in the `CTxMemPoolEntry` itself?  Then we would be reduced to just looping over the inputs of the incoming transactions and counting, for each distinct parent, how many of the dust outputs are spent. This is probably not simple to do at the moment (since in the package case, we don't have CTxMemPoolEntry objects to work on yet), but I have some refactors in mind for mempool acceptance which would make this easier to consider in the future.

Anyway, it might be nice to add a micro-benchmark for transaction validation of a single transaction with 1000 inputs, which all come from a single parent with 1000 outputs, just to make sure we're not slowing that case down inordinately.

instagibbs replied:
> we don't have CTxMemPoolEntry objects to work on yet

I don't think that's true? `PreChecks` has built a mempool entry for each package tx. We could track the exact outpoint that is dust in the mempool entry. I'll give that a shot.

instagibbs replied:
@sdaftuar this should do it if you think something like this is reasonable: https://github.com/instagibbs/bitcoin/commit/6007146613fe663f0af30c24c47a19792f8e691d

sdaftuar replied:
Hm, I I think that approach uses an extra 24 bytes per mempool entry, which strikes me as a bit heavy for what we need.  Let's see if this is fast enough as-is and if so, call it a day.

instagibbs replied:
Alternative is to store a bool, and use that to filter 99.9% of transactions that won't have dust, and then eat the O(num_outputs) cost to find the dust.

With 1000 inputs/outputs and a small number of dust outputs, I get:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           60,600.53 |           16,501.51 |    0.7% |      0.01 | `MempoolCheckEphemeralSpends`
```

When I make all the outputs dust(which can't happen currently due to IsStandard):
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           75,160.69 |           13,304.83 |    0.6% |      1.07 | `MempoolCheckEphemeralSpends`

```

number of outputs/inputs can be varied in the benchmark via `-asymptote=X`

sdaftuar replied:
I think this should be fine for performance.  Large transactions take orders of magnitude more time to validate than what is being added here (and for small transactions this is extremely fast).

murchandamus commented on src/rpc/mining.cpp:
```diff
@@ -493,7 +493,15 @@ static RPCHelpMan prioritisetransaction()
+    // Non-0 fees are not allowed for entry, and modification not allowed afterwards
+    const auto& tx = mempool.get(hash);
+    if (tx && std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, mempool.m_opts.dust_relay_feerate); })) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Priority is not supported for transactions with ephemeral dust.");
```
It seems to me that this would apply to any transaction with a dust output rather than just one that fulfills the criteria of "ephemeral dust". 

```suggestion
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Priority is not supported for transactions with dust outputs.");
```

instagibbs replied:
done

murchandamus commented on doc/release-notes-30239.md:
```diff
@@ -0,0 +1,12 @@
+-----------------------
+
+Ephemeral dust is a new concept that allows a single
+dust output in a tranasction, provided the transaction
```
```suggestion
dust output in a transaction, provided the transaction
```

instagibbs replied:
done

murchandamus reviewed (commented):
Concept ACK

instagibbs committed:
bench: Add basic CheckEphemeralSpends benchmark

sdaftuar commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,212 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+        bool should_rbf_eph_spend = outpoint_to_rbf && fuzzed_data_provider.ConsumeBool();
+
+        // Make small packages
+        const auto num_txs = should_rbf_eph_spend ? 1 : (size_t) fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
```
If it's not too hard, I was thinking it would be nice to have this fuzz test cover the package RBF case as well.

instagibbs replied:
I haven't looked lately but these harnesses should trigger package RBF behavior. I added the package mempool results checking to check the related invariants.

sdaftuar commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+
+    // Tests with parents in mempool
+
+    // Empty ancestors means this should never fail since it only checks parents' dust
```
Comment needs to be updated I think?

instagibbs replied:
updated

sdaftuar reviewed (commented):
Looks pretty good, I think most of my comments are nitty.  Along those lines, you may want to retouch the commit message of fcbfed4630bdee6c52644c89e972ba7c130f47d5, as I think there's a missing paren.

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,212 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+// Scan mempool for a tx that has spent dust, and return a "sibling"
+// spend of that dust to double-spend that dust-spend out of the mempool.
+// This assumes CheckMempoolEphemeralInvariants has passed for tx_pool.
+std::optional<COutPoint> GetEphemeralSibling(const CTxMemPool& tx_pool)
```
yeah that's confusing, reworded

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,479 @@
+        self.nodes[0].sendrawtransaction(cancel_sweep["hex"])
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"] for dusty_tx in dusty_txs] + [cancel_sweep["tx"]])
+
+        # Sweeps all dust, where most are already in-mempool
```
correct, updated comment

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,479 @@
+        # Doesn't spend in-mempool dust output from parent
+        unspent_sweep_tx = self.wallet.create_self_transfer_multi(fee_per_output=2000, utxos_to_spend=[dusty_tx["new_utxos"][0]], version=3)
+        assert_greater_than(unspent_sweep_tx["fee"], sweep_tx["fee"])
+        res = self.nodes[0].submitpackage([dusty_tx["hex"], unspent_sweep_tx["hex"]])
```
done

instagibbs commented on src/validation.cpp:
```diff
@@ -1592,6 +1610,19 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
+    // Now that we've bounded the resulting possible ancestry count, check parents for dust spends
+    if (m_pool.m_opts.require_standard) {
+        if (const auto ephemeral_violation{CheckEphemeralSpends(txns, m_pool.m_opts.dust_relay_feerate, m_pool)}) {
+            const Txid& parent_txid = ephemeral_violation.value();
```
out of date logging, updated

sdaftuar commented:
Code review ACK e01bb3a06693efc77a173afdf25ed7ba631316a2

instagibbs commented:
rebased on master due to conflict

theStack reviewed (commented):
Concept ACK

Mostly looked and reasoned about the core commit 92d7ad2b135668cb7f6b0fd32be15fd6dd2f915c so far. From what I understand the "only one dust output" limit would not be strictly necessary and is far less important than the other two rules (must be 0-fee, child must spend all dust from the parent) in order to disisincentivize creating dust UTXOs, or am I missing something?

What I left below are largely nits and follow-up ideas, feel free to ignore. Still want to look deeper into unit tests and the fuzzer.

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,74 @@
+{
+    // We never want to give incentives to mine this transaction alone
+    if ((base_fee != 0 || mod_fee != 0) &&
+        std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_fee); })) {
```
Added `HasDust` to make it more obvious where these checks are being done

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -491,7 +491,15 @@ static RPCHelpMan prioritisetransaction()
-    EnsureAnyMemPool(request.context).PrioritiseTransaction(hash, nAmount);
+    CTxMemPool& mempool = EnsureAnyMemPool(request.context);
+
+    // Non-0 fees are not allowed for entry, and modification not allowed afterwards
```
done

instagibbs commented on test/functional/test_framework/mempool_util.py:
```diff
@@ -21,6 +21,20 @@
 
 ORPHAN_TX_EXPIRE_TIME = 1200
 
+def assert_mempool_contents(test_framework, node, expected=None, sync=True):
```
Deferring to Future Work

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,480 @@
+
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+        dusty_tx = self.wallet.create_self_transfer_multi(fee_per_output=0, version=3)
+        self.add_output_to_create_multi_result(dusty_tx)
+
+        sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=dusty_tx["new_utxos"], version=3)
```
deferring to Future Work

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,480 @@
+        assert not test_res[0]["allowed"]
+        assert_equal(test_res[0]["reject-reason"], "dust")
+        # Reset priority
+        self.nodes[0].prioritisetransaction(txid=dusty_tx["txid"], dummy=0, fee_delta=-COIN)
```
solid idea, done

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,74 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include<policy/ephemeral_policy.h>
+#include<policy/policy.h>
```
fixed

instagibbs commented on test/functional/mempool_dust.py:
```diff
@@ -71,9 +71,39 @@ def test_dust_output(self, node: TestNode, dust_relay_fee: Decimal,
+        # Double dust, both unspent, with fees. Would have failed individual checks.
+        # Dust is 1 satoshi create_self_transfer_multi disallows 0
+        dusty_tx = self.wallet.create_self_transfer_multi(fee_per_output=1000, amount_per_output=1, num_outputs=2)
+        dust_txid = self.nodes[0].rpc.sendrawtransaction(hexstring=dusty_tx["hex"], maxfeerate=0)
```
fixed

instagibbs commented on doc/release-notes-30239.md:
```diff
@@ -0,0 +1,12 @@
+
+In other words, this type of transaction
+should be created in a transaction package where
+the dust is both created and spend simultaneously.
```
done

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,74 @@
+            // We look for an in-package or in-mempool dependency
+            CTransactionRef parent_ref = nullptr;
+            if (map_txid_ref.contains(parent_txid)) {
+                parent_ref = map_txid_ref[parent_txid];
```
taken

sdaftuar commented:
ACK 51c2394eb8158113e73659f4ae65c237813bd5d2

instagibbs commented on src/test/util/txmempool.cpp:
```diff
@@ -141,6 +141,54 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
+
+        Assert(dust_indexes.size() < 2);
+
+        if (dust_indexes.empty()) return;
```
Oof, this was making it a lot harder to hit the various cases! Fixed!

theStack reviewed (approved):
ACK 131bed19bdfc922328cad9781fa9122b6810a8fd

instagibbs commented on src/test/util/txmempool.cpp:
```diff
@@ -141,7 +141,7 @@ std::optional<std::string> CheckPackageMempoolAcceptResult(const Package& txns,
 }
 
-std::vector<uint32_t> GetDustIndexes(const CTransactionRef tx_ref, CFeeRate dust_relay_rate)
+std::vector<uint32_t> GetDustIndexes(const CTransactionRef& tx_ref, CFeeRate dust_relay_rate)
```
that's a mistake, I'll fix if I touch the PR

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,55 @@
+ * TxC, spends TxA's dust
+ *
+ * All the dust is spent if TxA+TxB+TxC is accepted, but the mining template may just pick
+ * up TxA+TxB rather than the three "legal configurations:
```
will fixup if I end up touching things

sdaftuar commented:
ACK 131bed19bdfc922328cad9781fa9122b6810a8fd

rkrux commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+        std::unordered_set<Txid, SaltedTxidHasher> processed_parent_set;
+        std::unordered_set<COutPoint, SaltedOutpointHasher> unspent_parent_dust;
+
+        for (const auto& tx_input : tx->vin) {
+            const Txid& parent_txid{tx_input.prevout.hash};
```
> Must be called for each transaction(package) if any dust is in the package. Checks that each transaction's parents have their dust spent by the child,

Should this loop run only for dust inputs by adding a `IsDust()` right at the beginning of this loop? I'm assuming this function is not responsible for the standard transaction validity checks as per the comment in the header file.

rkrux replied:
Aah nvm, `IsDust` works on TxOuts.

rkrux reviewed (commented):
Successful make and functional tests at 131bed19bdfc922328cad9781fa9122b6810a8fd.

I've gone through only the src/policy/*, src/rpc/*, src/validation.cpp yet. My comments are mostly around code structure. I plan to review the functional tests and src/tests/* very soon.

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        self.wallet.rescan_utxos()
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+        # Other topology tests require relaxation of submitpackage topology
```
I think "other topology" is a little under-specified here, could be helpful to name what tests you think could be added

glozow commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        self.sync_all()
+
+    # N.B. this extra_args can be removed post cluster mempool
+    def test_free_relay(self):
```
nit: `test_no_minrelay_fee` could be better

glozow commented on src/rpc/mining.cpp:
```diff
@@ -491,7 +492,15 @@ static RPCHelpMan prioritisetransaction()
+    // Non-0 fee dust transactions are not allowed for entry, and modification not allowed afterwards
+    const auto& tx = mempool.get(hash);
+    if (tx && HasDust(tx, mempool.m_opts.dust_relay_feerate)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Priority is not supported for transactions with dust outputs.");
```
Should this be added to release notes for RPC? I think the only way it'd be relevant to existing usage is in a acceptnonstdtxn=1 and non-0 dustrelayfee situation. But worth noting imo.

instagibbs replied:
alternatively could just check `mempool.m_opts.require_standard` as well?

glozow replied:
Yeah, adding `mempool.m_opts.require_standard` to the if condition would make sense - `acceptnonstdtxn` should turn off all dust checks including this one imo

instagibbs replied:
will add in follow-up

glozow reviewed (commented):
code review ack. I still need to review tests more thoroughly. None of these comments are blocking.

Trying to summarize why these would propagate via 1p1c package relay, please correct me if I'm wrong:

- Essentially there is no longer anything "wrong" with a transaction that has 1 dust output in it since `IsStandardTx` has been changed. It's not `TX_NOT_STANDARD`. It needs to be 0 fee to pass `CheckValidEphemeralTx`, and then it should fail on feerate for `TX_RECONSIDERABLE`. That's why the parent can be picked up by the package relay logic. Child also needs to pass `CheckEphemeralSpends`, but otherwise neatly fits into 1p1c.
- It's nice that `CheckValidEphemeralTx` comes before `TX_RECONSIDERABLE`, so we don't waste 1p1c cycles. But it's not strictly necessary to ensure an ephemeral tx is eligible for reconsideration.

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    // Add first grandparent to mempool and fetch entry
+    pool.addUnchecked(entry.FromTx(grandparent_tx_1));
+
+    // Ignores ancestors that aren't direct parents
```
maybe not the most important check, just makes sure that being an ancestor in mempool alone somehow isn't enough to get checked for dust

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    // Spend dust from one but not another is ok, as long as second grandparent has no child
+    BOOST_CHECK(!CheckEphemeralSpends({grandparent_tx_1, grandparent_tx_2, dust_spend}, minrelay, pool).has_value());
+
+    auto dust_non_spend_both_parents = make_tx({COutPoint{dust_txid, dust_index}, COutPoint{dust_txid_2, dust_index - 1}}, /*version=*/2);
```
I think it's correct, spending dust from first dusty tx but not second

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        # Sweeps all dust, where all dusty txs are already in-mempool
+        sweep_tx = self.wallet.create_self_transfer_multi(fee_per_output=25000, utxos_to_spend=all_parent_utxos, version=2)
+
+        res = self.nodes[0].submitpackage([dusty_tx["hex"] for dusty_tx in dusty_txs] + [sweep_tx["hex"]])
+        assert_equal(res['package_msg'], "success")
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"] for dusty_tx in dusty_txs] + [sweep_tx["tx"], cancel_sweep["tx"]])
```
could be worth a comment

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+#include <policy/ephemeral_policy.h>
+#include <policy/policy.h>
+
+bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate)
```
leaving as-is

instagibbs replied:
will handle on follow-up

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+                for (uint32_t out_index = 0; out_index < parent_ref->vout.size(); out_index++) {
+                    const auto& tx_output = parent_ref->vout[out_index];
+                    if (IsDust(tx_output, dust_relay_rate)) {
+                        unspent_parent_dust.insert(COutPoint(parent_txid, out_index));
+                    }
```
I don't think it would make the logic simpler or significantly more performant, and would rather the `IsStandardTx` check not be so tightly bound.

instagibbs commented on src/policy/policy.cpp:
```diff
@@ -142,11 +143,16 @@ bool IsStandardTx(const CTransaction& tx, const std::optional<unsigned>& max_dat
         }
     }
 
+    // Only MAX_DUST_OUTPUTS_PER_TX dust is permitted(on otherwise valid ephemeral dust)
+    if (num_dust_outputs > MAX_DUST_OUTPUTS_PER_TX) {
+        reason = "dust";
+        return false;
+    }
```
yes, `IsStandardTx` check is very early in PreChecks, not requiring access to any utxo information. including fees.

instagibbs commented on src/policy/policy.h:
```diff
@@ -77,6 +77,10 @@ static const unsigned int MAX_OP_RETURN_RELAY = 83;
  */
 static constexpr unsigned int EXTRA_DESCENDANT_TX_SIZE_LIMIT{10000};
 
+/**
+ * Maximum number of ephemeral dust outputs allowed.
+ */
+static constexpr unsigned int MAX_DUST_OUTPUTS_PER_TX{1};
```
Can change to `MAX_EPHEMERAL_DUST_OUTPUTS_PER_TX` if I touch things

instagibbs commented on src/rpc/mining.cpp:
```diff
@@ -491,7 +492,15 @@ static RPCHelpMan prioritisetransaction()
+    // Non-0 fee dust transactions are not allowed for entry, and modification not allowed afterwards
+    const auto& tx = mempool.get(hash);
+    if (tx && HasDust(tx, mempool.m_opts.dust_relay_feerate)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Priority is not supported for transactions with dust outputs.");
```
This is as intended. We don't allow priority on anything with dust, no matter how many non-zero dust outputs.

hodlinator commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        self.add_output_to_create_multi_result(dusty_tx)
+        assert_raises_rpc_error(-26, "min relay fee not met", self.nodes[0].sendrawtransaction, dusty_tx["hex"])
+
+        block_res = self.nodes[0].rpc.generateblock(self.wallet.get_address(), [dusty_tx["hex"]])
+        self.nodes[0].invalidateblock(block_res["hash"])
```
nit: Could validate/document that tx made it into the block to be sure, before invalidating.
```suggestion
        block_res = self.nodes[0].rpc.generateblock(self.wallet.get_address(), [dusty_tx["hex"]])
        assert_equal(self.nodes[0].getrawtransaction(dusty_tx["txid"], blockhash=block_res["hash"]), dusty_tx["hex"])
        self.nodes[0].invalidateblock(block_res["hash"])
```

instagibbs replied:
mmm, seems like a test of generateblock to me? going to leave as is

hodlinator reviewed (commented):
Code reviewed up until most of aa7c90a91adcf8ad9d6b94d5d17797cfd0eb5228.

Disclaimer: I posses far from full context when it comes to validation code, but at least as far as I can tell we are only modifying policy, not consensus.

PR title should probably have "p2p:" or some [other prefix](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#creating-the-pull-request).

stickies-v commented on src/rpc/mining.cpp:
```diff
@@ -491,7 +492,15 @@ static RPCHelpMan prioritisetransaction()
+
+    // Non-0 fee dust transactions are not allowed for entry, and modification not allowed afterwards
+    const auto& tx = mempool.get(hash);
+    if (tx && HasDust(tx, mempool.m_opts.dust_relay_feerate)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Priority is not supported for transactions with dust outputs.");
+    }
```
I independently came to the same conclusion, moving my review comment here:

I feel like d147d929f5093f71c39cb7efbb0dd3888f6c8114 could just be dropped entirely? If a miner has a reason to prioritize a transaction, I don't see why they'd be okay with that not being possible just because it has a dust output? So since this is a mining RPC, I feel like practically speaking they'd just patch it out, and there'd be no real use case left for this exception, so let's just save everyone involved the hassle?

instagibbs replied:
The same argument should be made for not allowing modified fees too right? There's an argument to be made, but I think it's more expansive than just the RPC.

hodlinator replied:
Not sure what you mean, seems you are arguing in the reverse direction or talking past?

I'm thinking the main use-case is for a miner or pool to increase priority of a transaction due to them getting paid out-of-band. While that is bad from a centralization aspect, having such a facility in unpatched versions of Bitcoin Core means miners/pools have an easier time keeping their nodes up-to-date (less moving parts), meaning they are more responsive to our releases.

instagibbs replied:
> Not sure what you mean, seems you are arguing in the reverse direction or talking past?

There are two places where we disallow modified fees to take effect:

1) prioritisetransaction, when the transaction is already in the mempool
2) on entry to mempool, in PreChecks

This ensures the invariant that unspent dust doesn't enter the utxo set ever (modulo reducing minrelay and block mining fee to 0).

We're assuming node operators/miners care about dust. If they do not, they can already set their dust relay rate to 0 and avoid all of these mechanisms entirely, ephemeral dust becomes a no-op.

If we end up being wrong and miners are ok putting one dust output in the utxo set but not multiple or whatever, we can revisit?

hodlinator replied:
Thanks for clarifying when modified fees are ignored.

> We're assuming node operators/miners care about dust. If they do not, they can already set their dust relay rate to 0 and avoid all of these mechanisms entirely, ephemeral dust becomes a no-op. (e.g., MARA reportedly doesn't enforce dust limits and runs an accelerator already)
>
> If we end up being wrong and miners are ok putting one dust output in the utxo set but not multiple or whatever, we can revisit?

My assumption would be that pools are fine with helping limit dust by default, but if they get paid through their accelerator service for a specific tx, any concern over dust goes out the window. Fighting an uncooperative Bitcoin Core after they already got paid out-of-band would annoy them.

Maybe it results in them changing their configs to no longer limit dust for *any transactions* when they otherwise would have.

stickies-v reviewed (commented):
Concept ACK. I've mostly reviewed the non-test code so far. No comments are blocking, I appreciate on a PR of this scope we can't address everything and things may need to happen in a follow-up.

hodlinator reviewed (commented):
Concept ACK 131bed19bdfc922328cad9781fa9122b6810a8fd

Thanks for working on this! Seems like an obvious win for L2s.

Haven't yet absorbed the full context sufficiently to fully A-C-K it myself. Hopefully my comments here and above provide value even though most are surface level. My personal priority with this review is to get it in better shape for others so they can A-C-K more easily if they agree.

tdb3 reviewed (commented):
Concept ACK

Great policy update that will provide flexibility to L2s.
Left a few comments, and want to take a deeper look at test cases.

glozow commented on src/bench/mempool_ephemeral_spends.cpp:
```diff
@@ -0,0 +1,83 @@
+
+    bench.run([&]() NO_THREAD_SAFETY_ANALYSIS {
+
+        CheckEphemeralSpends({tx2_r}, /*dust_relay_rate=*/CFeeRate(iteration * COIN / 10), pool);
```
Did you consider benching with the parent in package? I suppose it's about the same.

glozow reviewed (commented):
ACK 131bed19bdfc922328cad9781fa9122b6810a8fd

~Please lmk what you want to do with the remaining comments, I'm happy to reack or merge with a followup lined up.~
Happy to reack after rebase.

glozow commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+    // And set minrelay to 0 to allow ephemeral parent tx even with non-TRUC
+    mempool_opts.min_relay_feerate = CFeeRate(0);
+
+    // Don't worry about signaling replacement
+    mempool_opts.full_rbf = true;
```
Ah RIP, this is a merge conflict with #30592. That's the CI error.

instagibbs commented:
![image](https://github.com/user-attachments/assets/5b57516d-5ad8-4921-af05-71395e90b8e5)

rebased due to conflict with merge of https://github.com/bitcoin/bitcoin/pull/30592

going to go over all outstanding comments

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+                }
+            }
+
+            processed_parent_set.insert(parent_txid);
```
If it's not in mempool or the package, then it can't have unconfirmed dust. Is there a scenario you're concerned about?

tdb3 replied:
Was just focused on handling edge cases (and preventing harder to notice issues later).  Not sure if there's a scenario involving orphaning that might play out (haven't thought through all scenarios).

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+    return std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_rate); });
+}
+
+bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
+{
+    // We never want to give incentives to mine this transaction alone
+    if ((base_fee != 0 || mod_fee != 0) && HasDust(tx, dust_relay_rate)) {
+        return state.Invalid(TxValidationResult::TX_NOT_STANDARD, "dust", "tx with dust output must be 0-fee");
+    }
+
+    return true;
+}
```
good idea, will rename in follow-up

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
+{
+    // We never want to give incentives to mine this transaction alone
+    if ((base_fee != 0 || mod_fee != 0) && HasDust(tx, dust_relay_rate)) {
```
think we're sticking with 0-fee both base and modified

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+    return std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_rate); });
+}
+
+bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state)
```
will touch on follow-up

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+
+bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate)
+{
+    return std::any_of(tx->vout.cbegin(), tx->vout.cend(), [&](const auto& output) { return IsDust(output, dust_relay_rate); });
```
will touch on followup

instagibbs commented on src/policy/ephemeral_policy.cpp:
```diff
@@ -0,0 +1,78 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <policy/ephemeral_policy.h>
```
will touch in follow-up

instagibbs commented on src/bench/mempool_ephemeral_spends.cpp:
```diff
@@ -0,0 +1,83 @@
+
+    bench.run([&]() NO_THREAD_SAFETY_ANALYSIS {
+
+        CheckEphemeralSpends({tx2_r}, /*dust_relay_rate=*/CFeeRate(iteration * COIN / 10), pool);
```
It shouldn't effect the complexity much, as the cost is scanning all the parents' outputs

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,55 @@
+ * are the only way to bring fees.
+ */
+
+/** Returns true if transaction contains dust */
+bool HasDust(const CTransactionRef& tx, CFeeRate dust_relay_rate);
+
+/* All the following checks are only called if standardness rules are being applied. */
+
+/** Must be called for each transaction once transaction fees are known.
+ * Does context-less checks about a single transaction.
+ * Returns false if the fee is non-zero and dust exists, populating state. True otherwise.
+ */
+bool CheckValidEphemeralTx(const CTransactionRef& tx, CFeeRate dust_relay_rate, CAmount base_fee, CAmount mod_fee, TxValidationState& state);
+
+/** Must be called for each transaction(package) if any dust is in the package.
+ *  Checks that each transaction's parents have their dust spent by the child,
+ *  where parents are either in the mempool or in the package itself.
+ *  The function returns std::nullopt if all dust is properly spent, or the txid of the violating child spend.
+ */
+std::optional<Txid> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate, const CTxMemPool& tx_pool);
```
will touch in follow-up

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,55 @@
+/** Must be called for each transaction(package) if any dust is in the package.
+ *  Checks that each transaction's parents have their dust spent by the child,
+ *  where parents are either in the mempool or in the package itself.
+ *  The function returns std::nullopt if all dust is properly spent, or the txid of the violating child spend.
```
I'll consider the other comment's suggestions, but this really is just a belt and suspenders, the caller is given no guarantees on if you're calling with nullptrs.

instagibbs commented on test/functional/mempool_dust.py:
```diff
@@ -71,9 +71,39 @@ def test_dust_output(self, node: TestNode, dust_relay_fee: Decimal,
+        assert_equal(self.nodes[0].getrawmempool(), [dust_txid])
+
+        # Spends one dust along with fee input, leave other dust unspent to check ephemeral dust checks aren't being enforced
+        sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=[self.wallet.get_utxo(), dusty_tx["new_utxos"][0]])
```
good point, will take a look in followup

instagibbs commented on src/bench/mempool_ephemeral_spends.cpp:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    // Tx with many outputs
+    CMutableTransaction tx1 = CMutableTransaction();
```
maybe touch in follow-up

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    auto dust_non_spend = make_tx({COutPoint{dust_txid, dust_index - 1}}, /*version=*/2);
+
+    // Child spending non-dust only from parent should be disallowed even if dust otherwise spent
+    BOOST_CHECK(CheckEphemeralSpends({grandparent_tx_1, dust_non_spend, dust_spend}, minrelay, pool).has_value());
```
will touch in follow-up

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    // We first start with nothing "in the mempool", using package checks
+
+    // Trivial single transaction with no dust
+    BOOST_CHECK(!CheckEphemeralSpends({dust_spend}, minrelay, pool).has_value());
```
will touch in follow-up

instagibbs commented on src/test/txvalidation_tests.cpp:
```diff
@@ -89,6 +90,125 @@ static inline CTransactionRef make_tx(const std::vector<COutPoint>& inputs, int3
+    }
+    for (auto i{0}; i < 3; ++i) {
+        mtx.vout[i].scriptPubKey = CScript() << OP_TRUE;
+        mtx.vout[i].nValue = (i == 2) ? 0 : 10000;
```
will do in follow-up

instagibbs commented on src/test/util/txmempool.h:
```diff
@@ -47,6 +47,18 @@ std::optional<std::string>  CheckPackageMempoolAcceptResult(const Package& txns,
+/** Return indexes of the transaction's outputs that are considered dust
+ * at given dust_relay_rate.
+*/
+std::vector<uint32_t> GetDustIndexes(const CTransactionRef& tx_ref, CFeeRate dust_relay_rate);
```
was a history error, not touching

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+                                   PickValue(fuzzed_data_provider, mempool_outpoints).hash;
+            const auto delta = fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(-50 * COIN, +50 * COIN);
+            // We only prioritise out of mempool transactions since PrioritiseTransaction doesn't
+            // filter for ephemeral dust GetEntry
```
will remove in follow up

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+
+                CAmount amount_in{0};
+                for (int i = 0; i < num_in; ++i) {
+                    // Pop random outpoint
```
will add in followup

instagibbs commented on src/policy/ephemeral_policy.h:
```diff
@@ -0,0 +1,55 @@
+ *  where parents are either in the mempool or in the package itself.
+ *  The function returns std::nullopt if all dust is properly spent, or the txid of the violating child spend.
+ */
+std::optional<Txid> CheckEphemeralSpends(const Package& package, CFeeRate dust_relay_rate, const CTxMemPool& tx_pool);
```
will take a crack at this in follow-up

instagibbs replied:
will take a crack at it in follow-up

instagibbs replied:
pushed a lightly modified version that also reports the child txid on failure, rather than guessing it's the ultimate tx in the package. thanks! https://github.com/bitcoin/bitcoin/pull/31279

instagibbs commented on test/functional/mempool_dust.py:
```diff
@@ -71,9 +71,39 @@ def test_dust_output(self, node: TestNode, dust_relay_fee: Decimal,
+
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+        # Double dust, both unspent, with fees. Would have failed individual checks.
+        # Dust is 1 satoshi create_self_transfer_multi disallows 0
```
will fixup in follwup

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        self.nodes[0].invalidateblock(block_res["hash"])
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"]], sync=False)
+
+        # Also should happen if dust is swept
```
Yeah old language, will touch in follow-up

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+
+    chainstate.SetMempool(&tx_pool);
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 300)
```
good question, I'll take a look on follow-up

instagibbs replied:
> 11:24 <@dergoegge> I've suggested `LIMITED_WHILE(provider.remaining_bytes() > 0, iter_limit)` in the past as well
11:24 <@dergoegge> I think the goal is to let the fuzzer "choose" the number of iterations and both the above and consumebool achieve that
11:24 <@dergoegge> Iirc there were concerns around using remaining_bytes as really large inputs might cause timeouts
11:24 <@dergoegge> which is something we've seen with consumebool as well if the iteration limit is too high

I'm going to leave as-is in follow-up unless told otherwise

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+        }
+
+        // Remember all added transactions
+        std::set<CTransactionRef> added;
+        auto txr = std::make_shared<TransactionsDelta>(added);
+        node.validation_signals->RegisterSharedValidationInterface(txr);
```
used to have a use, let me look at it in follow-up

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+                const auto num_in = should_rbf_eph_spend ? 2 :
+                    last_tx ? fuzzed_data_provider.ConsumeIntegralInRange<int>(package_outpoints.size()/2 + 1, package_outpoints.size()) :
+                    fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
+                auto num_out = should_rbf_eph_spend ? 1 : fuzzed_data_provider.ConsumeIntegralInRange<int>(1, 4);
```
will take a look in follow-up

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+            // Create transaction to add to the mempool
+            const CTransactionRef tx = [&] {
+                CMutableTransaction tx_mut;
+                tx_mut.version = CTransaction::CURRENT_VERSION;
```
the version doesn't really matter since we set `mempool_opts.min_relay_feerate = CFeeRate(0);`

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+        std::vector<CTransactionRef> txs;
+
+        // Find something we may want to double-spend with two input single tx
+        std::optional<COutPoint> outpoint_to_rbf{GetChildEvictingPrevout(tx_pool)};
+        bool should_rbf_eph_spend = outpoint_to_rbf && fuzzed_data_provider.ConsumeBool();
```
will look in follow-up

instagibbs commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -137,6 +137,222 @@ std::unique_ptr<CTxMemPool> MakeMempool(FuzzedDataProvider& fuzzed_data_provider
+            // Last transaction in a package needs to be a child of parents to get further in validation
+            // so the last transaction to be generated(in a >1 package) must spend all package-made outputs
+            // Note that this test currently only spends package outputs in last transaction.
+            bool last_tx = num_txs > 1 && txs.size() == num_txs - 1;
```
will take a look in followup

instagibbs commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+
+        assert_equal(self.nodes[0].getrawmempool(), [])
+
+        dusty_tx = self.wallet.create_self_transfer_multi(
+            fee_per_output=0,
+            version=3
+        )
```
lots of code history blown away I'm sure :)

instagibbs commented:
addrman test failure for win64, assuming unrelated: https://github.com/bitcoin/bitcoin/pull/30239/checks?check_run_id=32868463159

fanquake commented:
> addrman test failure for win64, assuming unrelated: 

Yea this is a Wine issue (see #31071 & the linked issues). I restarted the job.

theStack reviewed (approved):
re-ACK 5c2e291060cca3be500f3af0f6f2d3fd2177a7c9

hodlinator commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+
+        sweep_tx = self.wallet.create_self_transfer_multi(utxos_to_spend=dusty_tx["new_utxos"], version=3)
+
+        self.nodes[0].prioritisetransaction(txid=dusty_tx["txid"], dummy=0, fee_delta=1000)
+        self.nodes[1].prioritisetransaction(txid=dusty_tx["txid"], dummy=0, fee_delta=1000)
```
Ah, totally missed that. Cheers!

Not sure what they add, it's not like there other transactions in the way. I guess they test that nothing funky is going on with the prioritization-logic resulting in dusty transactions being accepted when they shouldn't be.

hodlinator commented on test/functional/mempool_ephemeral_dust.py:
```diff
@@ -0,0 +1,482 @@
+        assert_equal(res['package_msg'], "success")
+        assert_mempool_contents(self, self.nodes[0], expected=[dusty_tx["tx"] for dusty_tx in dusty_txs] + [sweep_tx["tx"], cancel_sweep["tx"]])
+
+        self.generate(self.nodes[0], 25)
```
Seems unnecessarily wasteful for all future CPU cycles consumed by this test.

instagibbs replied:
pushed on follow-up
